pkg/file/xar.go

```diff
@@ -0,0 +1,132 @@
+// xar contains utilities to parse xar files, most of the logic here is a
+// simplified version extracted from the logic to sign xar files in
+// https://github.com/sassoftware/relic
+package file
+
+import (
+	"bytes"
+	"compress/zlib"
+	"crypto"
+	"encoding/binary"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+)
+
+// xarMagic is the [file signature][1] (or magic bytes) for xar
+//
+// [1]: https://en.wikipedia.org/wiki/List_of_file_signatures
+const xarMagic = 0x78617221
+
+const (
+	hashNone uint32 = iota
+	hashSHA1
+	hashMD5
+	hashSHA256
+	hashSHA512
+)
+
+var (
+	// ErrInvalidType is used to signal that the provided package can't be
+	// parsed because is an invalid file type.
+	ErrInvalidType = errors.New("invalid file type")
+	// ErrNotSigned is used to signal that the provided package doesn't
+	// contain a signature.
+	ErrNotSigned = errors.New("file is not signed")
+)
+
+type xarHeader struct {
+	Magic            uint32
+	HeaderSize       uint16
+	Version          uint16
+	CompressedSize   int64
+	UncompressedSize int64
+	HashType         uint32
+}
+
+type tocXar struct {
+	TOC toc `xml:"toc"`
+}
+
+type toc struct {
+	Signature  *any `xml:"signature"`
+	XSignature *any `xml:"x-signature"`
+}
+
+// CheckPKGSignature checks if the provided bytes correspond to a signed pkg
+// (xar) file.
+//
+// - If the file is not xar, it returns a ErrInvalidType error
+// - If the file is not signed, it returns a ErrNotSigned error
+func CheckPKGSignature(pkg io.Reader) error {
+	buff := bytes.NewBuffer(nil)
+	if _, err := io.Copy(buff, pkg); err != nil {
+		return err
+	}
+	r := bytes.NewReader(buff.Bytes())
```

Comment:
Nit: `*bytes.Buffer` (which `buff` is) implements `io.Reader`, so `bytes.NewReader is unnecessary (i.e. passing `buff` instead of `r` below would work just fine).

Reply:
Oh! Sorry about that, did not realize it.